File: leetcode/547.2.cpp
class Solution {
public:
	int findCircleNum(vector<vector<int>>& M) {
		int ret = 0;
		int size = M.size();
		vector<bool> mark(size, false);

		for (int i = 0; i < size; i++) {
			if (mark[i] == false) {
				bfs(i, mark, M);
				ret++;
			}
		}
		return ret;
	}

	void bfs(int s, vector<bool> &mark, vector<vector<int>>& M) {
		queue<int> q;
		q.push(s);
		mark[s] = true;
		while (!q.empty()) {
			s = q.front();
			// mark[cur] = true;
			q.pop();
			for (int i = 0; i < M.size(); i++) {
				if (M[s][i] == 1 && mark[i] == false) {
					mark[i] = true;
					q.push(i);
				}
			}
		}
		return;
	}

};

File: leetcode/547.2.java
class Solution {
    public int findCircleNum(int[][] M) {
        int ret = 0;
        int size = M.length;
        Boolean[] mark = new Boolean[size];
        for(int i=0; i<size; i++){
            mark[i] = false;
        }

        for(int i=0; i<size; i++){
            if(mark[i] == false){
                bfs(i, mark, M);
                ret++;
            }
        }
        return ret;
    }

    void bfs(int s, Boolean[] mark, int[][] M){
        Queue<Integer> q = new LinkedList<>();
        q.add(s);
        mark[s] = true;
        while(0 != q.size()){
            s = q.remove();
            // mark[cur] = true;
            for(int i=0; i<M.length; i++){
                if(M[s][i] == 1 && mark[i] ==false) {
                    mark[i] = true;
                    q.add(i);
                }
            }
        }
        return;
    }
}